### PR TITLE
[TK] Fix gather indices

### PR DESCRIPTION
**Pull request by Hardcode84**
Use original `memory` symbolic shape to determine dimensions access order.

### CODE REVIEW

**Comment by Hardcode84** [comment]
@raikonenfnu I still need to refactor indices calculation into proper place, but I will do it in separate PR from bugfix

File: shark_turbine/kernel/wave/codegen.py [hunk 1]
```diff
@@ -264,15 +264,27 @@ def handle_read(emitter: WaveEmitter, node: fx.Node):
...
+        # Apply input order to input indices, e.g. if original input_mapping was
+        # {M: iter(0), N: iter(1)} and source mem has shape (N, M), result will
+        # be (iter(1), iter(0))
+        input_mapping = mapping.map_input_indices(mem_shape)
 
         iters = mapping.iters
+
+        # As we only support identity output mapping for now, we can directly
+        # substitute iterators with corresponding expanded index.
         subs = [(sym, expr.start) for sym, expr in zip(iters.keys(), index.values())]
```
Comment (harsh-nod) [inline shark_turbine/kernel/wave/codegen.py:281]: Is there an assumption here about the ordering of the dimensions in index and iters? Say for example index was {K: ... , M: ... , N: ...} vs {M: .., N: ..., K: ...}, would that create a problem?

Comment (Hardcode84) [reply shark_turbine/kernel/wave/codegen.py:281]: I'm using mapping output shape for indexing if mapping is present, and for now, assert `mapping.is_output_identity()` is enforces output mapping is always `(iter(0), iter(1), ...)`

**Review by harsh-nod — APPROVED** [review]
thanks!

**Pull request merged**